File: problem/1047_RemoveAllAdjacentDuplicatesInString.js
// 1047. Remove All Adjacent Duplicates In String
// https://leetcode.com/problems/remove-all-adjacent-duplicates-in-string/
// You are given a string s consisting of lowercase English letters. A duplicate removal consists of choosing two adjacent and equal letters and removing them.

// We repeatedly make duplicate removals on s until we no longer can.

// Return the final string after all such duplicate removals have been made. It can be proven that the answer is unique.

/**
 * @param {string} s
 * @return {string}
 */
// string concat
var removeDuplicates = function(s) {
    let r = 0;
    while(r<s.length){
        if(s[r] ===s[r-1]){
            s = s.substring(0, r-1)+s.substring(r+1)
            r--
        } else {
            r++
        }
    }
    return s
};

// stack
var removeDuplicates = function(s) {
  let stack=[]
  for(let i=0; i<s.length; i++){
    if(s[i] !== stack[stack.length -1]){
      stack.push(s[i])
    } else {
      stack.pop()
    }
  }
  return stack.join('')
};

console.log(removeDuplicates("aaaaaaaa"))

File: c#-solution/1047. Remove All Adjacent Duplicates In String.cs
// 1047. Remove All Adjacent Duplicates In String
// https://leetcode.com/problems/remove-all-adjacent-duplicates-in-string/
// Tag: C#, Stack, String, Easy

public class Solution {
    public string RemoveDuplicates(string s)
    {
        var stack = new Stack<char>();
        foreach (var ch in s)
        {
            if (stack.Count > 0 && stack.Peek() == ch)
            {
                stack.Pop();
            }
            else
            {
                stack.Push(ch);
            }
        }
        return new string(stack.Reverse().ToArray());
    }
}

// string builder solution
    public string RemoveDuplicates(string s) {
        if(s.Length<2) return s;
        var str = new StringBuilder();
        for(var i =0; i<s.Length; i++){
            if(str.Length >0 && str[str.Length -1] == s[i]){
                str.Remove(str.Length -1, 1);
            }else{
                str.Append(s[i]);
            }
        }
          return str.ToString();
    }